### ISSUE TITLE
Docs: update all process code snippets in how-to and tutorial to literal include

### ISSUE THREAD

**Issue by sphuber** [aiidateam/aiida-core#4128]
This can only be done once `docs-revamp` is merged into `develop`, because the correct versions of those process plugins `add_multiply`, `ArithmeticAddCalculation` and `MultiplyAddWorkChain`.

**Comment by csadorf** [comment]
@sphuber Quick question, is this blocking to merge docs-revamp or for release?

**Comment by csadorf** [comment]
@sphuber This issue can now be resolved, correct?

**Comment by sphuber** [comment]
yes I am working on it

**Comment by sphuber** [comment]
I was able to apply the changes for the workflow how-to and tutorial without issue, however, the calculation how-to is written in such a way that just dropping in the code is a bit difficult. It uses slightly simplified snippets. Do you think it would be ok to have it add slightly more lines at the benefit of it showing code that is actually tested?

Another potential problem of literal including is that we now have to be very careful that if we update the code, we check the docs wherever we don't include the entire snippet but a range of lines. If the line numbers change, those snippets may go out of sync. Also the current calculation how-to relies heavily on line numbers, which I would have to adapt. I think that is possible in any case. Don't think it is necessary to reference line numbers to make it intelligible.

@csadorf and @ramirezfranciscof your input please.

**Comment by ramirezfranciscof** [comment]
I'm personally in favour of the how-tos being as complete as possible (here, in the sense of delivering working code). Could you maybe specify some of the problems you found when trying to use the snippets and how you were using them? Just to get a clearer idea of how are you expecting to use it, and also so I can test them in the same way.

Regarding the literal includes, I'm not sure what you mean. Are you talking about the calculation how-to as well here? The code snippets there are..."hardcoded" into the documentation (this is what "literal includes" mean?). I think this is ok because in no place we are saying "this is the actual code", but "this is a possible implementation", so there should be no expectation on these being up to date with the current internals. If we display the code that is actually in the source ("dinamically"?), then I think this will become harder to maintain as it is not just line numbers that might change, but the order of the commands, the names of the variables, etc. Maybe we could adapt the surrounding text to be more "general" at the cost of a bit less clarity, but I'm not sure what we are gaining by using the real code. Of course changes in the "interface" (let's say, chaning `spec` for `specifications`) would need to be manually updated, but this would also be the case if we were using the dynamical includes, unless you also were thinking of eliminating the explicit references to these critical objects from the text as well. Did I understood what you were proposing correctly? What would be the advantage of changing this?

**Comment by csadorf** [comment]
@sphuber If we are showing code examples that rely on a literalinclude, then a review of the related documentation will *always* be necessary, whether we rely on line numbers or not. So for me the question is rather, what benefit do we see in using a literalinclude in that case and how do we ensure that related documentation remain in sync?

**Comment by sphuber** [comment]
Thanks to both for the comments. To summarize the conundrum we are facing:

We either have code that is automatically tested, but it runs the risk of going out of sync with the text, or we remove the possibility but then the code snippets may be broken.

The latter used to be a big problem in the past when we were changing the API quite often and the documentation was always outdated with snippets that are broken, due to things as simple as broken imports. It is true that the chances of this happening after v1.0 are a lot smaller, however, having broken code in documentation snippets is pretty bad.

On the other hand, since the snippets we are showing are of classes that serve mostly for instructional purposes, the chances that they are changed are also slim. Not sure if there is a way that we can automatically check consistency or at the very least raise a warning when those files are touched to check the documentation

**Comment by ramirezfranciscof** [comment]
If this was just common code, maybe one "frankenstein-crazy" solution could be to have the code snippets in separate files all in a `docs/snippets` folder and include the relevant parts of those in the documentation while also being able to automatically run them as tests in a pre-commit step to check the integrity of the commands. Now, the extra complication that in order to test most of these plugin sections you need to actually register them, this I am not sure how to handle.

**Comment by csadorf** [comment]
@ramirezfranciscof I think that's the right solution and it's not "frankenstein-crazy" at all. Snippets that support the documentation should be maintained as part of the documentation (`docs/`) and those can be fully functional as well. We can control which parts are shown with the sphinx directive arguments. The function of these snippets is then automatically tested with each docs-build. Should there be any API issues, we discover them automatically and can update the snippets or roll-back accidental regressions.

**Comment by greschd** [comment]
Maybe you can use the `start_after` and `end_before` options in `include`, and put markers into the source files?

See reference here: https://docutils.sourceforge.io/docs/ref/rst/directives.html#include

Haven't tried this, but I imagine the source could look something like

```
<python code>
# START <SNIPPETLABEL>
<snippet code>
# END <SNIPPETLABEL>
<more python code>
```

and then you just use `START <SNIPPETLABEL>` and `END <SNIPPETLABEL>` as  `start_after` and `end_before`.

**Comment by sphuber** [comment]
> @ramirezfranciscof I think that's the right solution and it's not "frankenstein-crazy" at all. Snippets that support the documentation should be maintained as part of the documentation (`docs/`) and those can be fully functional as well. We can control which parts are shown with the sphinx directive arguments. The function of these snippets is then automatically tested with each docs-build. Should there be any API issues, we discover them automatically and can update the snippets or roll-back accidental regressions.

The problem with this that I see though, is that the snippets individually are not complete and so not executable. Only the entire class is, and since here we are talking about `CalcJob` and `Parser` instances, unit testing them are not trivial. What we will end up doing is more or less copy pasting the source code from the `aiida` package, copy testing the corresponding tests and then still we have to be careful when updating the code. I am not sure we will be making things simpler for us.

I think is important to realize that we are talking about code here that is explicitly added to `aiida-core` only for the purposes of testing and demonstration. I think this code should remain largely static and we won't be adding to it anytime soon. I am still leaning to avoiding all this duplication and be careful when those files are touched. I think we can make a pre-commit hook that when these files are touched there is a big warning to check them for consistency

**Comment by csadorf** [comment]
> I think is important to realize that we are talking about code here that is explicitly added to `aiida-core` only for the purposes of testing and demonstration.

Maybe we should rethink that?

**Comment by sphuber** [comment]
> Maybe we should rethink that?

Always open to reconsider it. But as it stands, it is very useful and we use it for quite a few purposes. If we were to get rid of it, we would have to solve those requirements differently. That is what I think makes this discussion a bit complicated: there are many things to consider and it is difficult to tackle the problems one by one in isolation.

**Comment by ramirezfranciscof** [comment]
> > I think is important to realize that we are talking about code here that is explicitly added to aiida-core only for the purposes of testing and demonstration.
>
> Maybe we should rethink that?

Uhm, sorry, I did not understand that. Rethink how? I don't think we have any other way of testing plugins and the plugin system other than having some actual plugin implemented due to this requirement of them being registered.

**Comment by ltalirz** [comment]
Note: I'm removing this from the "publish revised documentation milestone" (it was the only open issue and obviously the milestone has been reached).
It is still tracked in the "documentation restructure" project

**Comment by csadorf** [comment]
> > > I think is important to realize that we are talking about code here that is explicitly added to aiida-core only for the purposes of testing and demonstration.
> > 
> > 
> > Maybe we should rethink that?
> 
> Uhm, sorry, I did not understand that. Rethink how? I don't think we have any other way of testing plugins and the plugin system other than having some actual plugin implemented due to this requirement of them being registered.

But it does not have to be part of the core code base. It could just part of the documentation-related code base.

**Comment by sphuber** [comment]
> > Uhm, sorry, I did not understand that. Rethink how? I don't think we have any other way of testing plugins and the plugin system other than having some actual plugin implemented due to this requirement of them being registered.
> 
> But it does not have to be part of the core code base. It could just part of the documentation-related code base.

Yes that would be a possibility, were it not for the fact that part of the reasoning was to make it easy to load these plugins through the factories such that they can be used, in for example, the tutorials. This is already the case for @mbercx 's new tutorial. By having them in the `aiida-core` base, we can install them properly through entry points, making them easily loadable through the factories. They can of course be imported, but this is difficult to get right when submitting to the daemon when it comes to adding the sources to the Python path. Another advantage of having the sources in the core and installed with an entry point is that they become usable with utilities like `verdi plugin list`.

I will try to finish the changes in the doc using the code from the package and we see how it changes the current structure and see if that is acceptable to you guys.
